### DISCUSSION TITLE
Adding Type Annotations in patterns

### DISCUSSION THREAD

**Discussion by jrstrunk** [gleam-lang/gleam#3103] [Ideas & suggestions]
What does anyone think about adding the ability to add type annotations when pattern matching? An example of how this can be used straight from the radish package's decoder.gleam file is:
```gleam
use #(length: BitArray, rest) <- result.then(consume_till_crlf(rest, <<>>))
use length: String <- result.then(bit_array.to_string(length))
use length: Int <- result.then(int.parse(length))
``` 

This is currently a compile error since I cannot tell the compiler that I am expecting the `length` variable from the `consume_till_crlf` function to be a BitArray. This might be nice because currently if in my code I am expecting a BitArray value for the first element, but the `consume_till_crlf` function returns something else, say a `String`, then I will get an error when I call `bit_array.to_string`, even though that may be correct in the context of my application and it is the `consume_till_crlf` function that is returning the wrong type.

In the current version of Gleam, the above code snippet actually looks like this:
```gleam
use #(length, rest) <- result.then(consume_till_crlf(rest, <<>>))
use length <- result.then(bit_array.to_string(length))
use length <- result.then(int.parse(length))
```
*but could look like this:
```gleam
use #(length, rest) <- result.then(consume_till_crlf(rest, <<>>))
use length: String <- result.then(bit_array.to_string(length))
use length: Int <- result.then(int.parse(length))
```
Which in my opinion is a little less easy to read because it is not obvious at a glance why length is redeclared three times.

So, this addition could add:
1. The ability to tell the compiler what we are expecting when pattern matching on variables, giving us more accurate error messages if something is wrong
2. Better readability for variables being redeclared after a type change

Any thoughts?

**Comment by lpil** [comment]
Hello!

> *but could look like this:
>
> ```gleam
> use #(length, rest) <- result.then(consume_till_crlf(rest, <<>>))
> use length: String <- result.then(bit_array.to_string(length))
> use length: Int <- result.then(int.parse(length))
>```
> Which in my opinion is a little less easy to read because it is not obvious at a glance why length is redeclared three times.

This is valid code today.

The first example with the tuple annotation is written like so:
```gleam
use #(length, rest): #(BitArray, _) <- result.then(consume_till_crlf(rest, <<>>))
```

---

An aside: In Gleam it is called pattern matching rather than destructuring.

**Reply by jrstrunk** [reply]
Oh, that makes sense, thanks!

**Reply by lpil** [reply]
No prob!
